File: haskell/Day01.hs
module Day01 where
import Data.List (sort)
import Data.List.Split (splitOn)

parseLine :: String -> (Int, Int)
parseLine l = 
    let [x, y] = take 2 (splitOn "   " l)
    in (read x, read y)

parser :: String -> ([Int], [Int])
parser inp = sortInps (listAccum (map parseLine (splitOn "\n" inp)))

listAccum :: [(Int, Int)] -> ([Int], [Int])
listAccum pInp = foldl (\(fAcc, sAcc) (x, y) -> (x : fAcc, y : sAcc)) ([], []) pInp

sortInps :: ([Int], [Int]) -> ([Int], [Int])
sortInps (x,y) = (sort x, sort y)

processInps :: ([Int], [Int]) -> Int
processInps (x,y) = sum [abs (a - b)| (a,b) <- zip x y]

join :: Show a =>  String -> [a] -> String
join _ [] = ""
join delimiter (x:xs) = (show x) ++ delimiter ++ (join delimiter xs)

freqCounter :: [Int] -> [(Int, Int)]
freqCounter inp = map (\x -> ([head x], length x)) (group (sort inp))

part1 :: String -> String
part1 inp = show (processInps (parser inp))

solve :: String -> String
solve inp = part1 inp

File: python/day01.py
from collections import Counter


def parser(test: bool):
    fname = "test" if test else "inp"
    l_pairs = []
    r_pairs = []
    [[v := l.split("   "), l_pairs.append(int(v[0])), r_pairs.append(int(v[1]))] for l in open(
        f"./data/day01/{fname}.txt").read().splitlines()]
    return l_pairs, r_pairs


def part_1(test: bool = False) -> int:
    l_pairs, r_pairs = parser(test)
    l_pairs.sort()
    r_pairs.sort()
    summa = 0
    for x, y in zip(l_pairs, r_pairs):
        summa += abs(x - y)
    return summa


def part_2(test: bool = False) -> int:
    l_pairs, r_pairs = parser(test)

    r_ctr = dict(Counter(r_pairs))
    return sum(x * r_ctr.get(x, 0) for x in l_pairs)

if __name__ == "__main__":
    print(part_1())
    print(part_2())
